etna/datasets/hierarchical_structure.py

```diff
@@ -0,0 +1,179 @@
+        num_reachable_leafs: Dict[str, int] = dict()
+        for level in sorted(hierarchy_levels.keys(), reverse=True):
+            for node in hierarchy_levels[level]:
+                if node in self.level_structure:
+                    num_reachable_leafs[node] = sum(
+                        num_reachable_leafs[child_node] for child_node in self.level_structure[node]
+                    )
+
+                else:
+                    num_reachable_leafs[node] = 1
+
+        return num_reachable_leafs
+
+    def get_summing_matrix(self, target_level: str, source_level: str) -> scipy.sparse.base.spmatrix:
+        """Get summing matrix for transition from source level to target level.
+
+        Generation algorithm is based on summing matrix structure. Number of 1 in such matrices equals to
+        number of nodes on the source level. Each row of summing matrices has ones only for source level nodes that
+        belongs to subtree rooted from corresponding target level node. BFS order of nodes on levels view simplifies
+        algorithm to calculation necessary offsets for each row.
+
+        Parameters
+        ----------
+        target_level:
+            Name of target level.
+        source_level:
+            Name of source level.
+
+        Returns
+        -------
+        :
+            Summing matrix from source level to target level
+
+        """
+        try:
+            target_idx = self._level_to_index[target_level]
+            source_idx = self._level_to_index[source_level]
+        except KeyError as e:
+            raise ValueError("Invalid level name: " + e.args[0])
+
+        if target_idx >= source_idx:
+            raise ValueError("Target level must be higher in hierarchy than source level!")
+
+        target_level_segment = self.get_level_segments(target_level)
+        source_level_segment = self.get_level_segments(source_level)
+        summing_matrix = lil_matrix((len(target_level_segment), len(source_level_segment)))
+
+        current_source_segment_id = 0
+        for current_target_segment_id, segment in enumerate(target_level_segment):
+            num_reachable_leafs_left = self._segment_num_reachable_leafs[segment]
+
+            while num_reachable_leafs_left > 0:
+                source_segment = source_level_segment[current_source_segment_id]
+                num_reachable_leafs_left -= self._segment_num_reachable_leafs[source_segment]
+                summing_matrix[current_target_segment_id, current_source_segment_id] = 1
+                current_source_segment_id += 1
+
+        summing_matrix.tocsr()
+
+        return summing_matrix
+
+    def get_level_segments(self, level_name: str) -> List[str]:
+        """Get all segments from particular level."""
+        try:
+            return self._level_series[level_name]
+        except KeyError as e:
+            raise ValueError("Invalid level name: " + e.args[0])
+
+    def get_segment_level(self, segment: str) -> Union[str, None]:
```

Comment:
May be let's also fail here with exception "Segment {segment} is out of the hierarchy"